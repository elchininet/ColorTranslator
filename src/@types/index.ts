export type NumberOrString = number | string;

export interface HEXObject {
    r: string;
    g: string;
    b: string;
    a?: string;
}

export interface RGBObject {
    r: number;
    g: number;
    b: number;
    a?: number;
}

export interface RGBObjectGeneric {
    r: NumberOrString;
    g: NumberOrString;
    b: NumberOrString;
    a?: NumberOrString;
}

export interface HSLObject {
    h: number;
    s: number;
    l: number;
    a?: number;
}

export interface HSLObjectGeneric {
    h: number;
    s: NumberOrString;
    l: NumberOrString;
    a?: number;
}

export interface CMYKObject {
    c: number;
    m: number;
    y: number;
    k: number;
    a?: number;
}

export interface CMYKObjectGeneric {
    c: NumberOrString;
    m: NumberOrString;
    y: NumberOrString;
    k: NumberOrString;
    a?: number;
}

export interface RYBObject {
    r: number;
    y: number;
    b: number;
    a?: number;
}

export type RGYBObject = RGBObject | RYBObject;

export type Color = RGBObjectGeneric | HSLObjectGeneric | CMYKObjectGeneric;
export type ColorWithoutCMYK = RGBObjectGeneric | HSLObjectGeneric;

export type ColorInput = string | Color;
export type ColorInputWithoutCMYK = string | ColorWithoutCMYK;
export type HEXOutput = string | HEXObject;
export type RGBOutput = string | RGBObject;
export type HSLOutput = string | HSLObject;
export type CMYKOutput = string | CMYKObject;
export type ColorOutput = HEXOutput | RGBOutput | HSLOutput;

export interface ObjectProps<T> {
    [key: string]: T;
}

export interface Options {
    decimals: number;
    legacyCSS: boolean;
    spacesAfterCommas: boolean;
}

export type InputOptions = Partial<Options>;