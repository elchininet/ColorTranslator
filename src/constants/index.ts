export * from './numbers';
export * from './enums';
export * from './regexps';
export * from './errors';
