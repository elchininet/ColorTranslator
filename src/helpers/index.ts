import {
    Options,
    InputOptions,
    NumberOrString,
    ColorInput
} from '@types';
import {
    PCENT,
    HEX,
    MAX_DECIMALS,
    DEFAULT_OPTIONS,
    TypeOf
} from '#constants';

//---Has property
export const hasProp = <T>(obj: T, prop: string): boolean => Object.prototype.hasOwnProperty.call(obj, prop);

//---Get percent number
export const percentNumber = (percent: NumberOrString): number => {
    return +`${percent}`.replace(PCENT, '$1');
};

//---Get a percentage
export const percent = (percent: NumberOrString): number => PCENT.test(`${percent}`)
    ? percentNumber(percent)
    : Math.min(+percent, 100);

//---Convert to decimal
export const getDEC = (hex: string): number => {
    if (hex.length === 1) { hex += hex; }
    return parseInt(hex, 16);
};

//---Convert to hexadecimal
export const getHEX = (number: NumberOrString): string => {
    const hex = round(number, 0).toString(16).toUpperCase();
    if (hex.length === 1) { return `0x0${hex}`; }
    return `0x${hex}`;
};

//---Convert to final hexadecimal
export const toHEX = (h: NumberOrString): string => {
    let hex = round(h, 0).toString(16).toUpperCase();
    if (hex.length === 1) {
        hex = `0${hex}`;
    }
    return hex;
};

//---Calculate a decimal 255 from an RGB color
export const getBase255Number = (color: string, alpha = false): number => {
    if (!alpha && PCENT.test(color)) {
        return Math.min(255 * percentNumber(color) / 100, 255);
    }
    if (HEX.test(color)) {
        if (color.length === 3) {
            color += color.slice(-1);
        }
        return alpha
            ? round(color) / 255
            : round(color);
    }
    return Math.min(+color, alpha ? 1 : 255);
};

//---Calculate a decimal 0-1 value from CMYK value
export const getCMYKNumber = (color: string): number => Math.min(PCENT.test(color) ? percentNumber(color) / 100 : +color, 1);

//---Return an ordered string from an array values
export const getOrderedArrayString = (keys: string[]): string => [...keys].sort().join('').toUpperCase();

//---Round value
export const round = (value: NumberOrString, decimals = MAX_DECIMALS): number => {
    const exp = Math.pow(10, decimals);
    return Math.round(+value * exp) / exp;
};

//---Minimum and maximum
export const minmax = (n: number, min: number, max: number): number => Math.max(min, Math.min(n, max));

//--Radian to grades
export const grades = (radian: number): number => radian * 180 / Math.PI;

export const parseOptions = (options: Partial<Options>): Options => ({
    ...DEFAULT_OPTIONS,
    ...options
});

interface MatchOptions {
    legacyCSS: number;
    spacesAfterCommas: number;
}

export const getOptionsFromColorInput = (options: InputOptions, ...colors: ColorInput[]): Options => {
    const cssColors = colors.filter((color: ColorInput): boolean => typeof color === 'string') as string[];
    const matchOptions: MatchOptions = {
        legacyCSS: 0,
        spacesAfterCommas: 0
    };
    cssColors.forEach((color: string): void => {
        if (color.includes(',')){
            matchOptions.legacyCSS ++;
            const commasWithNextCharacter = color.match(/,( +|\d+)/g);
            if (
                new Set(commasWithNextCharacter).size === 1 &&
                / +/.test(commasWithNextCharacter[0].slice(1))
            ) {
                matchOptions.spacesAfterCommas ++;
            }
        }
    });
    return {
        decimals: typeof options.decimals === TypeOf.NUMBER
            ? options.decimals
            : DEFAULT_OPTIONS.decimals,
        legacyCSS: typeof options.legacyCSS === TypeOf.BOOLEAN
            ? options.legacyCSS
            : Boolean(
                cssColors.length &&
                matchOptions.legacyCSS === cssColors.length
            ) || DEFAULT_OPTIONS.legacyCSS,
        spacesAfterCommas: typeof options.spacesAfterCommas === TypeOf.BOOLEAN
            ? options.spacesAfterCommas
            : Boolean(
                cssColors.length &&
                matchOptions.spacesAfterCommas === cssColors.length
            ) || DEFAULT_OPTIONS.spacesAfterCommas
    };
};