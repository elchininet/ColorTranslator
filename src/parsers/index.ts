export { HEXStringParser } from './_HEXStringParser';
export { RGBStringParser } from './_RGBStringParser';
export { HSLStringParser } from './_HSLStringParser';
export { HWBStringParser } from './_HWBStringParser';
export { CIELabStringParser } from './_CIELabStringParser';
export { CMYKStringParser } from './_CMYKStringParser';