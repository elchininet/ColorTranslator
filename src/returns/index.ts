import {
    Color,
    ColorInput,
    ColorInputWithoutCMYK,
    ColorOutput,
    InputOptions,
    Options,
    RGBObject
} from '@types';
import {
    ColorModel,
    DEFAULT_BLEND_STEPS,
    Harmony,
    HarmonyString,
    Mix,
    MixString
} from '#constants';
import * as utils from '#color/utils';
import { getOptionsFromColorInput, getRGBObject } from '#color/extractors';

export const getColorReturn = <T>(
    color: ColorInput,
    model: ColorModel,
    decimals: number,
    translateFunction: (color: Color, decimals: number) => T
): T => {
    const rgbObject = getRGBObject(color, model);
    return translateFunction(rgbObject, decimals);
};

export const getBlendReturn = <T>(
    from: ColorInput,
    to: ColorInput,
    steps: number,
    decimals: number,
    translateFunction: (color: Color, decimals: number) => T
): T[] => {
    if (steps < 1) steps = DEFAULT_BLEND_STEPS;
    const fromRGBObject = getRGBObject(from);
    const toRGBObject = getRGBObject(to);
    const blendArray = utils.blend(fromRGBObject, toRGBObject, steps);
    return blendArray.map((color: RGBObject): T => {
        return translateFunction(color, decimals);
    });
};

interface GetBlendReturnObjectWithParameters<T>{
    from: ColorInput;
    to: ColorInput;
    thirdParameter?: number | InputOptions;
    fourthParameter?: InputOptions;
    translateFunction: (color: Color, decimals: number) => T;
}

interface GetBlendReturnStringWithParameters<T> extends GetBlendReturnObjectWithParameters<T> {
    cssFunction?: (color: T, options: Options) => string;
}

export function getBlendReturnWithParameters<T>(
    params: GetBlendReturnObjectWithParameters<T>
): T[];
export function getBlendReturnWithParameters<T>(
    params: GetBlendReturnStringWithParameters<T>
): string[];
export function getBlendReturnWithParameters<T>(
    params: GetBlendReturnStringWithParameters<T>
): T[] | string[] {
    const {
        from,
        to,
        thirdParameter,
        fourthParameter,
        translateFunction,
        cssFunction
    } = params;
    const result = typeof thirdParameter === 'number'
        ? getBlendReturn<T>(
            from,
            to,
            thirdParameter,
            fourthParameter?.decimals,
            translateFunction
        )
        : getBlendReturn<T>(
            from,
            to,
            DEFAULT_BLEND_STEPS,
            thirdParameter?.decimals,
            translateFunction
        );
    if (cssFunction) {
        return result.map((color: T): string => cssFunction(
            color,
            getOptionsFromColorInput(
                (
                    typeof thirdParameter === 'number'
                        ? fourthParameter
                        : thirdParameter
                ) ?? {},
                from,
                to
            )
        ));
    }
    return result;

}

interface GetMixReturn {
    colors: ColorInput[];
    secondParameter?: MixString | InputOptions;
    thirdParameter?: InputOptions;
    colorMixerFunction: (color: ColorInput[], mix: MixString, isCSS: boolean, options: Options) => string;
}

interface GetMixObjectReturn extends GetMixReturn {
    css: false;
}

interface GetMixCssReturn extends GetMixReturn {
    css: true;
}

export function getMixReturn<T>(params: GetMixObjectReturn): T;
export function getMixReturn(params: GetMixCssReturn): string;
export function getMixReturn<T>(
    params: GetMixObjectReturn | GetMixCssReturn
): T | string {
    const {
        colors,
        secondParameter,
        thirdParameter,
        colorMixerFunction,
        css
    } = params;
    const options = getOptionsFromColorInput(
        (
            typeof secondParameter === 'string'
                ? thirdParameter
                : secondParameter
        ) ?? {},
        ...colors
    );
    if (typeof secondParameter === 'string') {
        return colorMixerFunction(
            colors,
            secondParameter,
            css,
            options
        );
    }
    return colorMixerFunction(
        colors,
        Mix.ADDITIVE,
        css,
        options
    );
}

export const getHarmonyReturn = (
    harmony: HarmonyString,
    color: ColorInputWithoutCMYK,
    mode: MixString,
    options: Options
): ColorOutput[] => {
    return ({
        [Harmony.ANALOGOUS]:           utils.colorHarmony.buildHarmony(color, utils.analogous, mode, options),
        [Harmony.COMPLEMENTARY]:       utils.colorHarmony.buildHarmony(color, utils.complementary, mode, options),
        [Harmony.SPLIT_COMPLEMENTARY]: utils.colorHarmony.buildHarmony(color, utils.splitComplementary, mode, options),
        [Harmony.TRIADIC]:             utils.colorHarmony.buildHarmony(color, utils.triadic, mode, options),
        [Harmony.TETRADIC]:            utils.colorHarmony.buildHarmony(color, utils.tetradic, mode, options),
        [Harmony.SQUARE]:              utils.colorHarmony.buildHarmony(color, utils.square, mode, options)
    })[harmony];
};