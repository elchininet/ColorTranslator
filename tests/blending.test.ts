import { ColorTranslator } from '../src';

describe('ColorTranslator blending tests', (): void => {

    it('Blending deep equals', (): void => {

        const from = '#FF0000';
        const to = '#0000FF';

        const blendFunctions = [
            { colorFn: ColorTranslator.toHEX,  blendFn: ColorTranslator.getBlendHEX  },
            { colorFn: ColorTranslator.toHEXA, blendFn: ColorTranslator.getBlendHEXA },
            { colorFn: ColorTranslator.toRGB,  blendFn: ColorTranslator.getBlendRGB  },
            { colorFn: ColorTranslator.toRGBA, blendFn: ColorTranslator.getBlendRGBA },
            { colorFn: ColorTranslator.toHSL,  blendFn: ColorTranslator.getBlendHSL  },
            { colorFn: ColorTranslator.toHSLA, blendFn: ColorTranslator.getBlendHSLA }
        ];

        const result1 = [ '#FF0000', '#800080', '#0000FF' ];
        const result2 = [ '#FF0000', '#CC0033', '#990066', '#660099', '#3300CC', '#0000FF' ];
        const result3 = [ '#FF0000', '#BF0040', '#800080', '#4000BF', '#0000FF' ];

        blendFunctions.forEach((obj): void => {

            const r1 = result1.map(c => obj.colorFn(c));
            const r2 = result2.map(c => obj.colorFn(c));
            const r3 = result3.map(c => obj.colorFn(c));
            const r4 = result1.map(c => obj.colorFn(c, false));
            const r5 = result2.map(c => obj.colorFn(c, false));
            const r6 = result3.map(c => obj.colorFn(c, false));

            expect(obj.blendFn(from, to, r1.length)).toMatchObject(r1);
            expect(obj.blendFn(from, to, r2.length)).toMatchObject(r2);
            expect(obj.blendFn(from, to)).toMatchObject(r3);
            expect(obj.blendFn(from, to, r1.length, false)).toMatchObject(r4);
            expect(obj.blendFn(from, to, r2.length, false)).toMatchObject(r5);
            expect(obj.blendFn(from, to, 0, false)).toMatchObject(r6);

        });

        // Check default steps
        expect(ColorTranslator.getBlendHEX(from, to)).toMatchObject(ColorTranslator.getBlendHEX(from, to, 5));
        expect(ColorTranslator.getBlendHEX(from, to, -5)).toMatchObject(ColorTranslator.getBlendHEX(from, to, 5));

    });

});