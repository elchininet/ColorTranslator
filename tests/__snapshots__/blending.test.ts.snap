// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorTranslator blending tests Blending with decimals snapshots 1`] = `
[
  "rgb(255 0 0)",
  "rgb(191.25 0 63.75)",
  "rgb(127.5 0 127.5)",
  "rgb(63.75 0 191.25)",
  "rgb(0 0 255)",
]
`;

exports[`ColorTranslator blending tests Blending with decimals snapshots 2`] = `
[
  "rgb(255 0 0 / 1)",
  "rgb(191.25 0 63.75 / 1)",
  "rgb(127.5 0 127.5 / 1)",
  "rgb(63.75 0 191.25 / 1)",
  "rgb(0 0 255 / 1)",
]
`;

exports[`ColorTranslator blending tests Blending with decimals snapshots 3`] = `
[
  "hsl(0 100% 50%)",
  "hsl(340 100% 37.5%)",
  "hsl(300 100% 25%)",
  "hsl(260 100% 37.5%)",
  "hsl(240 100% 50%)",
]
`;

exports[`ColorTranslator blending tests Blending with decimals snapshots 4`] = `
[
  "hsl(0 100% 50% / 1)",
  "hsl(340 100% 37.5% / 1)",
  "hsl(300 100% 25% / 1)",
  "hsl(260 100% 37.5% / 1)",
  "hsl(240 100% 50% / 1)",
]
`;

exports[`ColorTranslator blending tests Blending with decimals snapshots 5`] = `
[
  {
    "b": 0,
    "g": 0,
    "r": 255,
  },
  {
    "b": 63.75,
    "g": 0,
    "r": 191.25,
  },
  {
    "b": 127.5,
    "g": 0,
    "r": 127.5,
  },
  {
    "b": 191.25,
    "g": 0,
    "r": 63.75,
  },
  {
    "b": 255,
    "g": 0,
    "r": 0,
  },
]
`;

exports[`ColorTranslator blending tests Blending with decimals snapshots 6`] = `
[
  {
    "a": 1,
    "b": 0,
    "g": 0,
    "r": 255,
  },
  {
    "a": 1,
    "b": 63.75,
    "g": 0,
    "r": 191.25,
  },
  {
    "a": 1,
    "b": 127.5,
    "g": 0,
    "r": 127.5,
  },
  {
    "a": 1,
    "b": 191.25,
    "g": 0,
    "r": 63.75,
  },
  {
    "a": 1,
    "b": 255,
    "g": 0,
    "r": 0,
  },
]
`;

exports[`ColorTranslator blending tests Blending with decimals snapshots 7`] = `
[
  {
    "h": 0,
    "l": 50,
    "s": 100,
  },
  {
    "h": 340,
    "l": 37.5,
    "s": 100,
  },
  {
    "h": 300,
    "l": 25,
    "s": 100,
  },
  {
    "h": 260,
    "l": 37.5,
    "s": 100,
  },
  {
    "h": 240,
    "l": 50,
    "s": 100,
  },
]
`;

exports[`ColorTranslator blending tests Blending with decimals snapshots 8`] = `
[
  {
    "a": 1,
    "h": 0,
    "l": 50,
    "s": 100,
  },
  {
    "a": 1,
    "h": 340,
    "l": 37.5,
    "s": 100,
  },
  {
    "a": 1,
    "h": 300,
    "l": 25,
    "s": 100,
  },
  {
    "a": 1,
    "h": 260,
    "l": 37.5,
    "s": 100,
  },
  {
    "a": 1,
    "h": 240,
    "l": 50,
    "s": 100,
  },
]
`;
